src/org/sosy_lab/java_smt/solvers/princess/PrincessModel.java

```diff
@@ -240,14 +253,76 @@ public String toString() {
   @Override
   public void close() {}
 
+  /** Tries to determine the Sort of a Term. */
+  private Sort getSort(IExpression pTerm) {
+    // Just using sortof() won't work as Princess may have simplified the original term
+    // FIXME: This may also affect other parts of the code that use sortof()
+    if (pTerm instanceof ITimes) {
+      ITimes times = (ITimes) pTerm;
+      return getSort(times.subterm());
+    } else if (pTerm instanceof IPlus) {
+      IPlus plus = (IPlus) pTerm;
+      return getSort(plus.apply(0));
+    } else if (pTerm instanceof IFormula) {
+      return creator.getBoolType();
+    } else {
+      // TODO: Do we need more cases?
+      return Sort$.MODULE$.sortOf((ITerm) pTerm);
+    }
+  }
+
+  /**
+   * Simplify rational values.
+   *
+   * <p>Rewrite <code>a/1</code> as <code>a</code>, otherwise return the term unchanged
+   */
+  private ITerm simplifyRational(ITerm pTerm) {
+    // TODO: Reduce the term further?
+    // TODO: Also do this for the values in the model?
+    if (Sort$.MODULE$.sortOf(pTerm).equals(creator.getRationalType())
+        && pTerm instanceof IFunApp
+        && ((IFunApp) pTerm).fun().name().equals("Rat_frac")
+        && pTerm.apply(1).equals(new IIntLit(IdealInt.ONE()))) {
+      return Rationals.int2ring((ITerm) pTerm.apply(0));
+    }
+    return pTerm;
+  }
+
   @Override
-  protected IExpression evalImpl(IExpression formula) {
-    IExpression evaluation = evaluate(formula);
-    if (evaluation == null) {
-      // fallback: try to simplify the query and evaluate again.
-      evaluation = evaluate(creator.getEnv().simplify(formula));
```

Comment:
What happens when `simplify()` is called for rationals?

Reply:
I've given it a try now, and unfortunately `simplify()` doesn't seem to solve the issue. The entire code here is meant to work around https://github.com/uuverifiers/princess/issues/7 and https://github.com/uuverifiers/princess/issues/8 in Princess and we might be able to simplify it again once these bugs have been closed.